raven/processes/wps_nalcms_zonal_stats_raster.py

```diff
@@ -0,0 +1,205 @@
+import json
+import logging
+import tempfile
+from collections import defaultdict
+from pathlib import Path
+
+from pywps import FORMATS, ComplexOutput, Process
+from pywps.inout.outputs import MetaFile, MetaLink4
+from rasterstats import zonal_stats
+from ravenpy.utilities.checks import single_file_check
+from ravenpy.utilities.geo import generic_vector_reproject
+from ravenpy.utilities.io import (
+    archive_sniffer,
+    crs_sniffer,
+    raster_datatype_sniffer,
+)
+
+from ..utils import (
+    NALCMS_PROJ4,
+    SIMPLE_CATEGORIES,
+    SUMMARY_ZONAL_STATS,
+    TRUE_CATEGORIES,
+    gather_dem_tile,
+    zonalstats_raster_file,
+)
+from . import wpsio as wio
+
+LOGGER = logging.getLogger("PYWPS")
+
+
+class NALCMSZonalStatisticsRasterProcess(Process):
+    """Given files containing vector data and raster data, perform zonal statistics of the overlapping regions"""
+
+    def __init__(self):
+        inputs = [
+            wio.shape,
+            wio.land_use_raster,
+            wio.simple_categories,
+            wio.raster_band,
+            wio.select_all_touching,
+        ]
+
+        outputs = [
+            wio.features,
+            wio.statistics,
+            ComplexOutput(
+                "raster",
+                "DEM grid subset by the requested shape.",
+                abstract="Zipped raster grid(s) of land-use using either standard or simplified UNFAO categories.",
+                as_reference=True,
+                supported_formats=[FORMATS.META4],
+            ),
+        ]
+
+        super(NALCMSZonalStatisticsRasterProcess, self).__init__(
+            self._handler,
+            identifier="nalcms-zonal-stats-raster",
+            title="NALCMS Land Use Zonal Statistics with raster output",
+            version="1.0",
+            abstract="Return zonal statistics, land-use cover, and raster grid for the CEC NALCMS based "
+            "on the boundaries of a vector file.",
+            metadata=[],
+            inputs=inputs,
+            outputs=outputs,
+            status_supported=True,
+            store_supported=True,
+        )
+
+    def _handler(self, request, response):
+
+        shape_url = request.inputs["shape"][0].file
+        simple_categories = request.inputs["simple_categories"][0].data
+        band = request.inputs["band"][0].data
+        touches = request.inputs["select_all_touching"][0].data
+
+        vectors = [".gml", ".shp", ".gpkg", ".geojson", ".json"]
+        vector_file = single_file_check(
+            archive_sniffer(shape_url, working_dir=self.workdir, extensions=vectors)
+        )
+        vec_crs = crs_sniffer(vector_file)
+
+        response.update_status("Accessed vector", status_percentage=5)
+
+        if (
+            "raster" in request.inputs
+        ):  # For raster files using the UNFAO Land Cover Classification System (19 types)
```

Comment:
Maybe put the comment above the line, to prevent Black from doing this ugly thing.